xdsl/transforms/reconcile_unrealized_casts.py

```diff
@@ -0,0 +1,107 @@
+import itertools
+from collections import deque
+from typing import Iterable
+from warnings import warn
+
+from xdsl.dialects import builtin
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext
+from xdsl.ir.core import Use
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import PatternRewriter
+
+
+def _try_remove_cast_chain(op: builtin.UnrealizedConversionCastOp):
+    # We remove several casts in a single pass, so we need to
+    # ignore casts that have already been removed earlier in the pass
+    # (i.e: no parent anymore).
+
+    if op.parent is not None:
```

Comment:
Will this ever fail? I don't think it actually can

Reply:
Thinking about it some more, I think modifying the IR structure while walking it is not allowed in python, by the contract of iterators. We should probably at least cache the operations we want to work on before calling this function, so we iterate over a list that is not mutated, even if the operations inside it are